langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Sort.java

```diff
@@ -68,15 +69,15 @@ public static ArrayValue sort(Strand strand, ArrayValue arr, Object direction, O
         Object[][] sortArr = new Object[arr.size()][2];
         Object[][] sortArrClone = new Object[arr.size()][2];
         if (function != null) {
-            BType elementType = ((BFunctionType) function.getType()).retType;
-            if (!(elementType.getTag() == TypeTags.UNION_TAG &&
-                    ((BUnionType) elementType).getMemberTypes().size() > 2)) {
-                elemType = elementType;
-            }
+            elemType = ((BFunctionType) function.getType()).retType;
             for (int i = 0; i < arr.size(); i++) {
                 sortArr[i][0] = function.call(new Object[]{strand, arr.get(i), true});
                 sortArr[i][1] = arr.get(i);
             }
+            if (elemType.getTag() == TypeTags.UNION_TAG &&
+                    ((BUnionType) elemType).getMemberTypes().size() > 2) {
+                elemType = TypeChecker.getType(sortArr[0][0]);
```

Comment:
Can you please explain what we're trying to do with this block? Is there a sample invocation?

Reply:
```
int[] arr = [1, 4, 2];
int[] sortedArr = arr.sort(array:ASCENDING, (i) => i);
```
when there is an arrow expression `((BFunctionType) function.getType()).retType` is a union type containing all ordered types. This is to cater to getting the type of the array members we need to sort in such a scenario.

Reply:
1. Shall we add a comment explaining why we're doing it? Also it's not just an arrow function right? There can even be a normal function with a compatible union type as the return type?
2. A union should have at least 2 members. If it has only 2, don't we have to change `elemType`?
3. Can't `sortArr` or `sortArr[0]` be empty here?

Reply:
1. Added a comment. That condition checking was only added for arrow functions. When it's an arrow function we definitely know that the function return type is a union with more than 2 member types.  In other cases if return type is a union it can contain maximum of two members. Thus this condition is checked `if (elemType.getTag() == TypeTags.UNION_TAG && ((BUnionType) elemType).getMemberTypes().size() > 2)` and the type of values in sortArr is taken.
2. This is already handled in https://github.com/ballerina-platform/ballerina-lang/blob/4fc699331c77169e5668cf013ce06ae69a572d84/langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Sort.java#L92
3. yes. therefore changed the position of the condition checking for arrow functions